xarray/core/weighted.py

```diff
@@ -0,0 +1,307 @@
+
+
+def _maybe_get_all_dims(
+    dims: Optional[Union[Hashable, Iterable[Hashable]]],
+    dims1: Tuple[Hashable, ...],
+    dims2: Tuple[Hashable, ...],
+):
+    """ the union of all dimensions
+
+    `dims=None` behaves differently in `dot` and `sum`, so we have to apply
+    `dot` over the union of the dimensions
+
+    """
+
+    if dims is None:
+        dims = tuple(sorted(set(dims1) | set(dims2)))
+
+    return dims
+
+
+def _sum_of_weights(
+    da: "DataArray",
+    weights: "DataArray",
+    dim: Optional[Union[Hashable, Iterable[Hashable]]] = None,
+    axis=None,
+) -> "DataArray":
+    """ Calcualte the sum of weights, accounting for missing values """
+
+    # we need to mask DATA values that are nan; else the weights are wrong
+    mask = where(da.notnull(), 1, 0)  # binary mask
+
+    # need to infer dims as we use `dot`
+    dims = _maybe_get_all_dims(dim, da.dims, weights.dims)
+
+    # use `dot` to avoid creating large da's
+    sum_of_weights = dot(mask, weights, dims=dims)
+
+    # find all weights that are valid (not 0)
+    valid_weights = sum_of_weights != 0.0
+
+    # set invalid weights to nan
+    return sum_of_weights.where(valid_weights)
+
+
+def _weighted_sum(
+    da: "DataArray",
+    weights: "DataArray",
+    dim: Optional[Union[Hashable, Iterable[Hashable]]] = None,
+    axis=None,
+    skipna: Optional[bool] = None,
+    **kwargs
+) -> "DataArray":
+    """Reduce a DataArray by a by a weighted `sum` along some dimension(s)."""
+
+    # need to infer dims as we use `dot`
+    dims = _maybe_get_all_dims(dim, da.dims, weights.dims)
+
+    # use `dot` to avoid creating large da's
+
+    # need to mask invalid DATA as dot does not implement skipna
+    if skipna or skipna is None:
+        return where(da.isnull(), 0.0, da).dot(weights, dims=dims)
+
+    return dot(da, weights, dims=dims)
+
+
+def _weighted_mean(
+    da: "DataArray",
+    weights: "DataArray",
+    dim: Optional[Union[Hashable, Iterable[Hashable]]] = None,
+    axis=None,
+    skipna: Optional[bool] = None,
+    **kwargs
+) -> "DataArray":
+    """Reduce a DataArray by a weighted `mean` along some dimension(s)."""
+
+    # get weighted sum
+    weighted_sum = _weighted_sum(
+        da, weights, dim=dim, axis=axis, skipna=skipna, **kwargs
+    )
+
+    # get the sum of weights
+    sum_of_weights = _sum_of_weights(da, weights, dim=dim, axis=axis)
+
+    # calculate weighted mean
+    return weighted_sum / sum_of_weights
+
+
+class Weighted:
+    """A object that implements weighted operations.
+
+    You should create a Weighted object by using the `DataArray.weighted` or
+    `Dataset.weighted` methods.
+
+    See Also
+    --------
+    Dataset.weighted
+    DataArray.weighted
+    """
+
+    __slots__ = ("obj", "weights")
+
+    @overload
+    def __init__(self, obj: "DataArray", weights: "DataArray") -> None:
+        ...
+
+    @overload
+    def __init__(self, obj: "Dataset", weights: "DataArray") -> None:
+        ...
```

Comment:
for typing, maybe remove as it breaks flake8 (F811)

Reply:
would then require removing typing from `__init__` as `Union["DataArray", "Dataset"]` makes mypy fail

Reply:
should actually be fixed on [pyflakes](https://github.com/PyCQA/pyflakes/issues/320)  and [flake8](https://gitlab.com/pycqa/flake8/issues/423) so azure probably has an old version installed, maybe:
``` python
def __init__(...) # noqa: F811 TODO: remove once pyflakes/ flake8 on azure is updated
```